src/chpasswd.c

```diff
@@ -601,8 +601,8 @@ int main (int argc, char **argv)
 				 * shadow, create one.
 				 */
 				newsp.sp_namp  = name;
-				/* newsp.sp_pwdp  = NULL; will be set later */
-				/* newsp.sp_lstchg= 0;    will be set later */
+				newsp.sp_pwdp  = NULL;
+				newsp.sp_lstchg= 0;
```

Comment:
Who uses those values?  I didn't find any code that reads them (although it goes through many layers, and I may have missed something).  I'd like to be sure that this is necessary.  Otherwise, we may be silencing some static analyzer by doing this initialization.

Please show a path that makes this initialization necessary.  Otherwise, I feel safer dropping this patch.

Reply:
Here's my opinion on initializing stuff to silence analyzers: <https://github.com/neomutt/neomutt/pull/4310#issuecomment-2110072727>.

Reply:
I didn't find them either. I'm fine with dropping them, but in general terms I prefer to initialize variables as they make the coder think what value they want in the variable instead of relying on the compiler to set some random value.

Reply:
The benefit of leaving the value undefined is that a static analyzer is able to analyze some bugs, but if you initialize the value, it's impossible.

See for example this program:
```c
$ cat bug2.c 
#include <stdlib.h>

void
g(int *x)
{
	if (arc4random() % 2)
		*x = 42;
}

int
main(void)
{
	int x;

	g(&x);
	return x;  // maybe uninitialized use
}
```

It has a bug, because there's one path where I'm reading the value without having assigned to the variable, so the contents are undefined.

A static analyzer is able to detect that there's a path that leads to Undefined Behavior, and warns about it:

```c
$ clang-tidy ./bug2.c 
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "./bug2.c"
No compilation database found in /home/alx/tmp/analyzer or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
1 warning generated.
/home/alx/tmp/analyzer/bug2.c:16:2: warning: Undefined or garbage value returned to caller [clang-analyzer-core.uninitialized.UndefReturn]
        return x;  // maybe uninitialized use
        ^      ~
/home/alx/tmp/analyzer/bug2.c:13:2: note: 'x' declared without an initial value
        int x;
        ^~~~~
/home/alx/tmp/analyzer/bug2.c:15:2: note: Calling 'g'
        g(&x);
        ^~~~~
/home/alx/tmp/analyzer/bug2.c:6:6: note: Assuming the condition is false
        if (arc4random() % 2)
            ^~~~~~~~~~~~~~~~
/home/alx/tmp/analyzer/bug2.c:6:2: note: Taking false branch
        if (arc4random() % 2)
        ^
/home/alx/tmp/analyzer/bug2.c:8:1: note: Returning without writing to '*x'
}
^
/home/alx/tmp/analyzer/bug2.c:15:2: note: Returning from 'g'
        g(&x);
        ^~~~~
/home/alx/tmp/analyzer/bug2.c:16:2: note: Undefined or garbage value returned to caller
        return x;  // maybe uninitialized use
        ^      ~
```

However, if you chose a value, the static analyzer has no way to know that you had a bug.  The bug is still there; just that it's different now.

```c
$ cat bug3.c 
#include <stdlib.h>

void
g(int *x)
{
	if (arc4random() % 2)
		*x = 42;
}

int
main(void)
{
	int x = 0;

	g(&x);
	return x;  // let's say a 0 should never reach here
}
```

```sh
$ clang-tidy bug3.c 
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "bug3.c"
No compilation database found in /home/alx/tmp/gcc or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
```

By initializing the value to 0, we've transformed an obvious bug into a silent bug.  Moreover, a maintainer 10 years from now, who doesn't know what the code is doing, will not know if that 0 is necessary, or if it was just a random value that the programmer put to silence a bad static analyzer.

But did anyone check if having a 0 results in good behavior?  Probably not, because it's a path that in theory should never happen.